File: loan-repayment-java/src/test/java/com/loan/repayment/service/RepaymentPlanServiceTest.java
package com.loan.repayment.service;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.MockitoAnnotations;

import com.loan.repayment.exception.ApplicationException;
import com.loan.repayment.factory.RepaymentPlanRequestFactory;

/**
 * 
 * @author Ramachandra
 *
 */
class RepaymentPlanServiceTest {

    @InjectMocks
    private RepaymentPlanService repaymentPlanService;

    @BeforeEach()
    public void init() {
        MockitoAnnotations.openMocks(this);
    }

    @Test
    void testRepaymentPlan() {
        final var banksResponse = repaymentPlanService
                .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequest());
        assertEquals(24, banksResponse.getBorrowerPayments().size(), "Expected the payments to be 24");
    }

    @Test
    void testRepaymentPlanWithBlankLoanAmount() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithBlankLoanAmount());
        });
        assertEquals("Loan Amount should not be blank to for Generating the Repayment Schedule", exception.getMessage(),
                "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithNullLoanAmount() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithNullLoanAmount());
        });
        System.out.println(exception.getMessage());
        assertEquals("Loan Amount should not be blank to for Generating the Repayment Schedule", exception.getMessage(),
                "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithZeroLoanAmount() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroLoanAmount());
        });
        assertEquals("Loan Amount should not be less than zero or zero for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithInvalidDoubleLoanAmount() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInvalidDoubleLoanAmount());
        });
        assertEquals("Loan Amount Zero not a valid double value to for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithNullNominalRate() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithNullNominalRate());
        });
        assertEquals("Nominal rate should not be blank to for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithBlankNominalRate() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithBlankNominalRate());
        });
        assertEquals("Nominal rate should not be blank to for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithZeroNominalRate() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroNominalRate());
        });
        assertEquals("Nominal rate should not be less than zero or zero for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithInvalidDoubleNominalRate() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService.repaymentPlan(
                    RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInvalidDoubleNominalRate());
        });
        assertEquals("Nominal rate Zero not a valid double value to for Generating the Repayment Schedule",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithInValidTenureWithLoanSchedule() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService.repaymentPlan(
                    RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInValidTenureWithLoanSchedule());
        });
        assertEquals(
                "In the duration 306 months the Repayment Schedule cannot be completed, Please provide the correct duration",
                exception.getMessage(), "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithZeroMonthsDuration() {
        ApplicationException exception = assertThrows(ApplicationException.class, () -> {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroMonthsDuration());
        });
        assertEquals("Duration of the Repayment Schedule should be minimum 1", exception.getMessage(),
                "Expected the error messages to match");
    }

    @Test
    void testRepaymentPlanWithOneMonthsDuration() {
        final var banksResponse = repaymentPlanService
                .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithOneMonthDuration());
        assertEquals(1, banksResponse.getBorrowerPayments().size(), "Expected the payments to be 1");
    }
}

File: loan-repayment-kotlin/src/test/kotlin/com/loan/repayment/service/RepaymentPlanServiceTest.kt
package com.loan.repayment.service

import com.loan.repayment.exception.ApplicationException
import com.loan.repayment.factory.RepaymentPlanRequestFactory
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.MockitoAnnotations

/**
 *
 * @author Rama chandra
 */
internal class RepaymentPlanServiceTest {
    @InjectMocks
    private lateinit var repaymentPlanService : RepaymentPlanService

    @BeforeEach
    fun init() {
        MockitoAnnotations.openMocks(this)
    }

    @Test
    fun testRepaymentPlan() {
        val banksResponse = repaymentPlanService
                .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequest())
        assertEquals(24 , banksResponse.borrowerPayments.size , "Expected the payments to be 24")
    }

    @Test
    fun testRepaymentPlanWithBlankLoanAmount() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithBlankLoanAmount())
        }
        assertEquals("Loan Amount should not be blank to for Generating the Repayment Schedule" , exception.message ,
                "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithZeroLoanAmount() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroLoanAmount())
        }
        assertEquals("Loan Amount should not be less than zero or zero for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithInvalidDoubleLoanAmount() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInvalidDoubleLoanAmount())
        }
        assertEquals("Loan Amount Zero not a valid double value to for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithNullNominalRate() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithNullNominalRate())
        }
        assertEquals("Nominal rate should not be blank to for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithBlankNominalRate() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithBlankNominalRate())
        }
        assertEquals("Nominal rate should not be blank to for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithZeroNominalRate() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService
                    .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroNominalRate())
        }
        assertEquals("Nominal rate should not be less than zero or zero for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithInvalidDoubleNominalRate() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService.repaymentPlan(
                    RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInvalidDoubleNominalRate())
        }
        assertEquals("Nominal rate Zero not a valid double value to for Generating the Repayment Schedule" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithInValidTenureWithLoanSchedule() {
        val exception = assertThrows(ApplicationException::class.java) {
            repaymentPlanService.repaymentPlan(
                    RepaymentPlanRequestFactory.createRepaymentPlanRequestWithInValidTenureWithLoanSchedule())
        }
        assertEquals(
                "In the duration 306 months the Repayment Schedule cannot be completed, Please provide the correct duration" ,
                exception.message , "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithZeroMonthsDuration() {
        val exception : ApplicationException = assertThrows(ApplicationException::class.java) {
            repaymentPlanService.repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithZeroMonthsDuration())
        }
        assertEquals("Duration of the Repayment Schedule should be minimum 1" , exception.message ,
                "Expected the error messages to match")
    }

    @Test
    fun testRepaymentPlanWithOneMonthsDuration() {
        val banksResponse = repaymentPlanService
                .repaymentPlan(RepaymentPlanRequestFactory.createRepaymentPlanRequestWithOneMonthDuration())
        assertEquals(1 , banksResponse.borrowerPayments.size ,
                "Expected the payments to be 1")
    }
}